From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Vincent Strubel <clipos@ssi.gouv.fr>
Date: Fri, 11 Jan 2013 18:36:58 +0100
Subject: [PATCH 11/32] clip: vserver: Map ipv4 addrs in ping socket bind()

Make sure that a ping source bind address is allowed in the nx context
according to the configuration (single source ip, loopback remapped etc)

Patch: 1511_vserver_ping_map_addr.patch
---
 net/ipv4/ping.c | 24 ++++++++++++++++++------
 1 file changed, 18 insertions(+), 6 deletions(-)

diff --git a/net/ipv4/ping.c b/net/ipv4/ping.c
index f4af4a5792b21..ec1b88c08ae94 100644
--- a/net/ipv4/ping.c
+++ b/net/ipv4/ping.c
@@ -375,19 +375,28 @@ static int ping_check_bind_addr(struct sock *sk, struct inet_sock *isk,
 	return 0;
 }
 
-static void ping_set_saddr(struct sock *sk, struct sockaddr *saddr)
+static int ping_set_saddr(struct sock *sk, struct sockaddr *saddr)
 {
+	int err;
+	struct inet_sock *isk = inet_sk(sk);
 	if (saddr->sa_family == AF_INET) {
-		struct inet_sock *isk = inet_sk(sk);
 		struct sockaddr_in *addr = (struct sockaddr_in *) saddr;
-		isk->inet_rcv_saddr = isk->inet_saddr = addr->sin_addr.s_addr;
+		struct nx_v4_sock_addr nsa;
+		err = v4_map_sock_addr(isk, addr, &nsa);
+		if (err)
+			return err;
+		v4_set_sock_addr(isk, &nsa);
 #if IS_ENABLED(CONFIG_IPV6)
 	} else if (saddr->sa_family == AF_INET6) {
 		struct sockaddr_in6 *addr = (struct sockaddr_in6 *) saddr;
-		struct ipv6_pinfo *np = inet6_sk(sk);
-		sk->sk_v6_rcv_saddr = np->saddr = addr->sin6_addr;
+		struct nx_v6_sock_addr nsa;
+		err = v6_map_sock_addr(inet, addr, &nsa);
+		if (err)
+			return err;
+		v6_set_sock_addr(inet, &nsa);
 #endif
 	}
+	return 0;
 }
 
 static void ping_clear_saddr(struct sock *sk, int dif)
@@ -426,8 +435,11 @@ int ping_bind(struct sock *sk, struct sockaddr *uaddr, int addr_len)
 	if (isk->inet_num != 0)
 		goto out;
 
+	err = ping_set_saddr(sk, uaddr);
+	if (err)
+		goto out;
+
 	err = -EADDRINUSE;
-	ping_set_saddr(sk, uaddr);
 	snum = ntohs(((struct sockaddr_in *)uaddr)->sin_port);
 	if (ping_get_port(sk, snum) != 0) {
 		ping_clear_saddr(sk, dif);
-- 
2.17.0

